### ISSUE TITLE
false positive MissingParamType when implement a abstract method of a trait

### ISSUE THREAD

**Issue by pfazzi** [vimeo/psalm#1513]
Consider the following code:
```php
/**
 * @template T
 */
trait ValueObjectTrait
{

    /**
     * @psalm-var T
     */
    protected $value;

    /**
     * @psalm-param T $value
     *
     * @param $value
     */
    private function setValue($value): void
    {
        $this->validate($value);

        $this->value = $value;
    }

    /**
     * @psalm-param T $value
     *
     * @param $value
     */
    abstract protected function validate($value): void;
}
```

```php
final class UniqueDocumentCode
{
    /**
     * @use ValueObjectTrait<string>
     */
    use ValueObjectTrait;

    protected function validate($value): void
    {
        Assertion::maxLength($value, 30);
    }
}
```
I get the following error:
```
INFO: MissingParamType - UniqueDocumentCode.php:15:33 - Parameter $value has no provided type protected function validate($value): void
```
But i have already declared the parameter type in ValueObjectTrait, so i should not get the notice.

**Comment by muglug** [comment]
Yeah, seems like trait inheritance isn't following the same rules as class inheritance ([which works](https://psalm.dev/r/6288be91bf))

**Comment by muglug** [comment]
There are also bugs if you do add a param: https://psalm.dev/r/00f3c9a0d4
